### ISSUE TITLE
Generic test framework support

### ISSUE THREAD

**Issue by SamMousa** [infection/infection#499]
# Introduction
I'm interested in Codeception support specifically and have read the related issues and PRs.

From what I can see adding a new framework is relatively complex and there's no real documentation on what's the minimum required.

Based on feedback in the issues and PRs in my view the minimal implementation would need to:
1. provide infection with a command to run tests
2. store the code coverage in a known location
3. provide infection with a way to know if tests failed
4. provide infection with a way to intercept the autoloader

To make running tests a bit more efficient:
- Tests should stop as soon as there is one failure

To provide more detailed reports:
- Tests should provide some output with respect to memory usage.

One of the comments mentioned the requirement for the test framework to support running specific tests. As far as I can tell this is not a strict requirement to doing mutation testing, but it is more of a performance optimization.

# Proposal

## MVP
I propose implementing a generic framework that uses simple and established approaches:
1. Configure a command to run tests.
2. Infection uses substitution to set a location for coverage, for example: `composer test --coverage {coveragefile}`.
3. The test command must use exit codes to indicate it's success.
4. The user configures a bootstrap file that's read by the test framework.
5. Infection uses substitution to set a location for the altered bootstrap file. This allows us to hook the autoloader in a generic manner.
## Efficiency
Since this is a configuration for the test framework it's up to the user to make his testing framework abort as soon as possible.

## Reporting
Based on a quick review of the source code, the only extra bit of information that can be passed back to infection is the memory usage. For the generic adapter we could simply use a regex configuration option that is used against the output.

# Goal
The goal of this proposal is to provide a quick way for people to get started with infection if they already have a test infrastructure in place.
I imagine that a PR for this functionality would be simple and therefore easier to review / merge.
It also provides a way for people on unsupported frameworks use infection testing without the upfront investment of adding support for their framework to infection.

**Comment by BackEndTea** [comment]
> One of the comments mentioned the requirement for the test framework to support running specific tests. As far as I can tell this is not a strict requirement to doing mutation testing, but it is more of a performance optimization.

Yes this is not a strict requirement, but it is **highly** encouraged to have the test framework support this. Its the difference between a 15 minute or 2 day mutation test run.

**Comment by SamMousa** [comment]
>   Its the difference between a 15 minute or 2 day mutation test run.

I'm hoping that's an exaggeration; if not we can define it to be mandatory ;-)

Could you point me to the interface that defines this? I assume infection somehow checks which tests cover which files?
The running a specific tests part is simply solved via another substitution; but this leaves identifying the name(s) for tests covering a specific file.

**Comment by sanmai** [comment]
Suppose you're running PHPUnit like so:
```
vendor/bin/phpunit --coverage-xml=coverage/coverage-xml --log-junit=coverage/phpunit.junit.xml
```
Now try:
```
grep -RC10 covered coverage/coverage-xml/
```
You would see something along the lines of:
```grep
Standard.php.xml-      <line nr="113">
Standard.php.xml:        <covered by="Tests\Pipeline\StandardTest::testPipelineInPipeline"/>
Standard.php.xml:        <covered by="Tests\Pipeline\StandardTest::testPipelineReadsFromPipeline"/>
Standard.php.xml-      </line>
```

So provided there are standard logs, you can know which tests covered which line. [Infection takes care of these decisions about here.](https://github.com/infection/infection/blob/master/src/TestFramework/Coverage/CodeCoverageData.php) And then, in case of PHPUnit, you can run only specific tests, and only.

In other words, the problem is two-fold:
1. Getting the right logs from a testing framework.
2. Getting it run only specific tests.

**Comment by theofidry** [comment]
Closing as a duplicate of #499

**Comment by maks-rafalko** [comment]
> Closing as a duplicate of #499

this PR is a #499 so it's not a duplicate of itself

**Comment by theofidry** [comment]
Corrected... Thanks
